File: splitsy-backend/src/domain/users/value_objects/password.py
import re
from dataclasses import dataclass, field

from src.domain.common.errors import ValidationError
from src.domain.common.value_objects import ValueObject

PASSWORD_REGEX = re.compile(r"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^A-Za-z0-9]).{8,}$")


@dataclass(frozen=True)
class PasswordProps:
    value: str = field(repr=False)


class Password(ValueObject[PasswordProps]):
    def __init__(self, props: PasswordProps) -> None:
        super().__init__(props)
        self._value = props.value

    def get_secret_value(self) -> str:
        return self._value

    @staticmethod
    def create(value: str) -> "Password":
        if not isinstance(value, str):
            raise ValidationError(f"Invalid format: {value}")

        if not PASSWORD_REGEX.match(value):
            raise ValidationError(
                "Must be at least 8 characters long and include "
                "uppercase, lowercase, number, and special character."
            )

        return Password(PasswordProps(value=value))

    def __str__(self) -> str:
        return "<hidden>"

File: splitsy-frontend/src/domain/shared/value-objects/Password.ts
import { ValueObject } from "./ValueObject";

interface PasswordAttributes {
  value: string;
}

const PASSWORD_REGEX = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[^A-Za-z0-9]).{8,}$/;

export class Password extends ValueObject<PasswordAttributes> {
  private constructor(props: PasswordAttributes) {
    super(props);
  }

  public static create(password: string): Password {
    if (!password || password.length === 0) {
      throw new Error('Password is required')
    }
    if (!PASSWORD_REGEX.test(password)) {
      throw new Error('Password must be at least 8 characters and include uppercase, lowercase, a number and a special character')
    }
    return new Password({ value: password })
  }

  public get value(): string {
    return this.props.value;
  }
}
